### ISSUE TITLE
Exception in test infrastructure - System.InvalidOperationException: Collection was modified after the enumerator was instantiated

### ISSUE THREAD

**Issue by 4creators** [dotnet/runtime#11063]
For the second time I have encountered the following error during CI OSX test run.

https://ci.dot.net/job/dotnet_coreclr/job/master/job/x64_checked_osx10.12_innerloop_tst_prtest/5550/

```
02:51:26   Unhandled Exception: System.InvalidOperationException: Collection was modified after the enumerator was instantiated.
02:51:26      at System.Collections.Generic.Stack`1.Enumerator.MoveNext()
02:51:26      at Xunit.Sdk.DisposalTracker.Dispose() in C:\projects\xunit\src\xunit.execution\Sdk\DisposalTracker.cs:line 26
02:51:26      at Xunit.Sdk.ExtensibilityPointFactory.Dispose() in C:\projects\xunit\src\xunit.execution\Sdk\ExtensibilityPointFactory.cs:line 54
02:51:26      at Xunit.Sdk.TestFramework.Dispose() in C:\projects\xunit\src\xunit.execution\Sdk\Frameworks\TestFramework.cs:line 49
02:51:26      at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
02:51:26   --- End of stack trace from previous location where exception was thrown ---
02:51:26      at System.Threading.ThreadPoolWorkQueue.Dispatch()
```

**Comment by safern** [comment]
Happened again in: https://dev.azure.com/dnceng/public/_build/results?buildId=646104&view=results

https://helix.dot.net/api/2019-06-17/jobs/6259c638-1c48-444a-9df5-b3d8b89354e4/workitems/PayloadGroup0/console

**Comment by Anipik** [comment]
happened again in https://github.com/dotnet/runtime/pull/43651

**Comment by ViktorHofer** [comment]
Probably coming from: https://github.com/xunit/xunit/blob/a8614d34999c889e1c8014f679de95d20eae1304/src/xunit.execution/Sdk/DisposalTracker.cs#L26-L27

I see @danmosemsft already linked to the xunit issue: https://github.com/xunit/xunit/issues/1855

**Comment by stephentoub** [comment]
I've been hitting this reasonably frequently, e.g.
https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-44409-merge-8a1a8bc2383a41f093/PayloadGroup0/console.e860a3c4.log?sv=2019-07-07&se=2020-11-29T17%3A18%3A38Z&sr=c&sp=rl&sig=0mS1HKsUz%2FgaWANt6pRwMCfnGXyaOIXCnwq327pVblg%3D
```
Unhandled exception.   Discovering: Exceptions.ForeignThread.XUnitWrapper
System.InvalidOperationException: Collection was modified after the enumerator was instantiated.
   at System.Collections.Generic.Stack`1.Enumerator.MoveNext()
   at Xunit.Sdk.DisposalTracker.Dispose() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\DisposalTracker.cs:line 26
   at Xunit.Sdk.TestFramework.Dispose() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\TestFramework.cs:line 52
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__140_1(Object state)
   at System.Threading.QueueUserWorkItemCallbackDefaultContext.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
Aborted (core dumped)
```

**Comment by stephentoub** [comment]
> I see @danmosemsft already linked to the xunit issue: xunit/xunit#1855

Which is closed.

**Comment by ViktorHofer** [comment]
> Which is closed.

As it was fixed in xunit v3 which hasn't shipped yet.

**Comment by naricc** [comment]
I am also seeing this intermittently on Mono runtime test lanes.

**Comment by radical** [comment]
Seeing this in https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-master-21906b723775461185/PayloadGroup0/console.81c9eaa1.log?sv=2019-07-07&se=2020-12-23T13%3A13%3A31Z&sr=c&sp=rl&sig=D65mDzIZzZ36F68UhRgc5vOffnOAUJ8%2BXHelojRWpIA%3D 

on [`Mono Linux x64 Release @ Ubuntu.1804.Amd64.Open`](https://dnceng.visualstudio.com/public/_build/results?buildId=907179&view=ms.vss-test-web.build-test-results-tab&runId=28934866&paneView=debug)

**Comment by ViktorHofer** [comment]
Happened again in https://dev.azure.com/dnceng/public/_build/results?buildId=932232&view=ms.vss-test-web.build-test-results-tab&runId=29520600&resultId=102132&paneView=attachments.

**Comment by kunalspathak** [comment]
I saw this in https://dev.azure.com/dnceng/public/_build/results?buildId=966315&view=results

**Comment by Anipik** [comment]
There has been an uptick in these failures.
https://dev.azure.com/dnceng/public/_build/results?buildId=997790&view=logs&j=e8cc9eb8-98f1-55f3-b6df-6c384ccb39b6&t=7ec1aa02-369c-51d1-7a60-8e709c4509a9
https://dev.azure.com/dnceng/public/_build/results?buildId=996779&view=logs&j=f5bbf73d-4d5d-5c21-6748-018818b1d8ec&t=49b7db0f-7ab2-56d3-55e4-df171e8b3dbb
https://dev.azure.com/dnceng/public/_build/results?buildId=998551&view=logs&j=e8cc9eb8-98f1-55f3-b6df-6c384ccb39b6

**Comment by ViktorHofer** [comment]
We would need to patch xunit.execution's library to mitigate the issue until xunit v3 is released. Maybe there is another hacky way that avoids reshipping that assembly ie via reflection?

**Comment by ViktorHofer** [comment]
@dougbu @jaredpar have you seen such issue in aspnetcore / roslyn?

**Comment by dougbu** [comment]
My https://runfo.azurewebsites.net -fu isn't the greatest but I couldn't find matches for Xunit.Sdk.DisposalTracker.Dispose in aspnetcore-ci pipeline over the last two weeks.

**Comment by jaredpar** [comment]
This is a good query for aspnetcore-ci 

https://runfo.azurewebsites.net/search/helix-logs/?bq=definition%3Aaspnetcore-ci+started%3A%7E6+result%3A%21succeeded&lq=logkind%3Aconsole+text%3A%22disposaltracker%22

Paging through I didn't see any of the failures. Same query will show failures for runtime.

**Comment by Anipik** [comment]
failed again here https://dev.azure.com/dnceng/public/_build/results?buildId=1009373&view=logs&j=2296eecc-f5ab-523f-c15d-f45eef1bfdea&t=d328b091-66a6-5d96-5a21-3cf77f57f009

**Comment by briansull** [comment]
Failed again here:
https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-49029-merge-e7a0fb9c237e44cdaf/PayloadGroup0/console.cf19b373.log?sv=2019-07-07&se=2021-03-23T23%3A31%3A42Z&sr=c&sp=rl&sig=FgM7e689DITmgVRp899rLb%2F%2FjB%2FVFzskAIwVSF6%2Ftqk%3D

**Comment by BruceForstall** [comment]
I'm seeing this a lot recently. E.g.,

https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-8dbe8f0829404ca684/PayloadGroup0/console.e7510d3c.log?sv=2019-07-07&se=2021-03-24T22%3A45%3A23Z&sr=c&sp=rl&sig=%2B7KpNEyyf2iw3AEOAjx6T2JdeP6x4per%2BQOqDE%2FvaPg%3D

https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-2101a5426fbf4350bc/PayloadGroup0/console.7eba733b.log?sv=2019-07-07&se=2021-03-24T23%3A16%3A19Z&sr=c&sp=rl&sig=vMLdKlvctkITNxsNM11WC3Xi1JvXzCeMKNrVWnbIIWM%3D

(from https://dev.azure.com/dnceng/public/_build/results?buildId=1023241&view=ms.vss-test-web.build-test-results-tab&runId=31829680&resultId=109719&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab)

**Comment by danmoseley** [comment]
Linking https://github.com/xunit/xunit/pull/1846

**Comment by echesakovMSFT** [comment]
I don't think xunit/xunit#1846 is a root cause here. I fixed that issue with https://github.com/dotnet/arcade/pull/2360 in our clone of xunit.console.dll runner a while ago. That issue was happening at the beginning (or sometimes in the middle) of tests execution during `AssemblyLoadContext.ResolveUsingResolvingEvent` while the one we are experiencing recently seems to be at the end of tests runs.

**Comment by jkoritzinsky** [comment]
This has started flaring up again. Not entirely sure why.

**Comment by ViktorHofer** [comment]
Another hit: https://dev.azure.com/dnceng/public/_build/results?buildId=1120636&view=ms.vss-test-web.build-test-results-tab&runId=34101518&resultId=102283&paneView=attachments.

**Comment by fanyang-mono** [comment]
I constantly see this issue for my PR to enabling all passing runtime tests on Android arm64 #49662

**Comment by BruceForstall** [comment]
I'm still seeing this commonly. E.g., 

CoreCLR windows x86 Checked jitstressregs3_x86_noavx

https://dev.azure.com/dnceng/public/_build/results?buildId=1196639&view=ms.vss-test-web.build-test-results-tab&runId=35883378&resultId=109760&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab

https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-c5eb8272aa734a85bd/PayloadGroup0/console.32ebffd4.log?sv=2019-07-07&se=2021-07-10T20%3A05%3A19Z&sr=c&sp=rl&sig=fx2Ka%2FE6CfFou4ZWSwxWcRKZjJ0wnVfzn%2Bq10kf13GY%3D

**Comment by BruceForstall** [comment]
cc @agocke (If you haven't seen this). I mentioned this to you recently.

**Comment by agocke** [comment]
Yup. unfortunately nothing we can do, AFAIK. XUnit 2 is not going to get a fix for this.

**Comment by BruceForstall** [comment]
Another: https://dev.azure.com/dnceng/public/_build/results?buildId=1196861&view=ms.vss-test-web.build-test-results-tab&runId=35887762&resultId=111100&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab

**Comment by BruceForstall** [comment]
Maybe we should fork xunit 2 (like we forked xunit.console?) and fix it? We've been dealing with this for years.

**Comment by BruceForstall** [comment]
Another: https://dev.azure.com/dnceng/public/_build/results?buildId=1198515&view=ms.vss-test-web.build-test-results-tab&runId=35924882&resultId=110557&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab

Another: https://dev.azure.com/dnceng/public/_build/results?buildId=1198515&view=ms.vss-test-web.build-test-results-tab&runId=35924818&resultId=109760&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab

**Comment by BruceForstall** [comment]
Another: https://dev.azure.com/dnceng/public/_build/results?buildId=1196639&view=ms.vss-test-web.build-test-results-tab&runId=35883378&resultId=109760&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab

**Comment by stephentoub** [comment]
@bradwilson, what would you recommend we do here?  The scale at which we run tests means we hit this failure in xunit frequently, causing PRs to fail at a somewhat regular frequency.

**Comment by BruceForstall** [comment]
Another: https://dev.azure.com/dnceng/public/_build/results?buildId=1198803&view=ms.vss-test-web.build-test-results-tab&runId=35934972&resultId=110576&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab

**Comment by BruceForstall** [comment]
Another: https://dev.azure.com/dnceng/public/_build/results?buildId=1208210&view=ms.vss-test-web.build-test-results-tab&runId=36130382&resultId=109767&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab

**Comment by BruceForstall** [comment]
Another: https://dev.azure.com/dnceng/public/_build/results?buildId=1208612&view=ms.vss-test-web.build-test-results-tab&runId=36138312&resultId=109767&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab

**Comment by bradwilson** [comment]
@stephentoub I have not changed my mind on shipping any fixes for v2, including this one. https://github.com/xunit/xunit/pull/1846#issuecomment-434126418

**Comment by stephentoub** [comment]
So just to make sure I understand the situation,  xunit v3 is nowhere in sight, xunit v2 will not see any fixes,  and anyone who needs an update is encouraged to fork and build (and,  if we were to do it,  likely publish for anyone else to consume) a custom version?

**Comment by bradwilson** [comment]
@stephentoub 

>xunit v3 is nowhere in sight

Work continues with alpha level CI builds available via MyGet.

>xunit v2 will not see any fixes

There are no plans at this time.

>anyone who needs an update is encouraged to fork and build

If they are unable to find a workaround, that is the best option.

>if we were to do it, likely publish for anyone else to consume

I can't speculate on what you would do.

>Is there anything we could do to help change that situation?

At this point in time, this project is a personal effort with very limited available time to invest (a small handful of hours per week). This is why I have pushed back on doing any 2.x work: if I went down that path, then there would be no time left to work on v3.

**Comment by BruceForstall** [comment]
Another: https://dev.azure.com/dnceng/public/_build/results?buildId=1232688&view=ms.vss-test-web.build-test-results-tab&runId=36700392&resultId=109744&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab

**Comment by danmoseley** [comment]
> This is why I have pushed back on doing any 2.x work: if I went down that path, then there would be no time left to work on v3.

I understand. If we provided and tested the fix entirely ourselves, how costly would it be for you to review it, commit it and push the output of CI to Nuget? This would be a binary with exactly the current content + this fix. It would be fine to push it as preview status, I think.

**Comment by clairernovotny** [comment]
One of the issues is that the CI has completely changed since then....not sure if the old CI even works anymore in the branch.

If @bradwilson is ok with it, and the team has a PR with a targeted fix, I could get a build pipeline working somewhere for publishing.

**Comment by sandreenko** [comment]
another: https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-55558-merge-9cb71a7742ec456497/PayloadGroup0/console.4d9565f3.log?sv=2019-07-07&se=2021-08-02T12%3A13%3A55Z&sr=c&sp=rl&sig=rNpgee4mr%2F3XFuv4x80LwIPvbOKPMvg7%2BHpxXzpn8%2FY%3D

**Comment by clairernovotny** [comment]
> @clairernovotny Claire Novotny FTE that would be great -- I'd like to take you up on that offer -- yes please.
> 
> We can then find someone to look into a fix or hackaround (it may not be immediate due to .NET 6 activities)

I just need the okay from @bradwilson

**Comment by bradwilson** [comment]
My problem is this: it sets a precedent that (potentially arbitrary) fixes and features will be accepted for v2 and given an official release. Who is going to be on the hook for reviewing, merging, and releasing all those things? Because it won't be me.

**Comment by stephentoub** [comment]
Again here:
https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-55575-merge-19f5f323fe664dfc88/Interop/console.e2e8004d.log?sv=2019-07-07&se=2021-08-02T18%3A24%3A31Z&sr=c&sp=rl&sig=pmXBY1jpNkRXtgD0nJcOlymmCiKgm9ZnNtZnAl91l6M%3D

**Comment by danmoseley** [comment]
> My problem is this: it sets a precedent that (potentially arbitrary) fixes and features will be accepted for v2 

@bradwilson it's your project -- you control it and my assumption is that you can simply continue to say no. 😺

**Comment by bradwilson** [comment]
Why not release your own private build?

**Comment by danmoseley** [comment]
I had (mis?) understood @clairernovotny to be offering to set up a CI to build a private build.

**Comment by clairernovotny** [comment]
I was offering to setup a real build in an servicing branch and get it published. My other interest is secure supply chain as I want it built with reproducible settings.

**Comment by bradwilson** [comment]
@danmoseley My objections here are all around releasing a publicly sanctioned release of xUnit.net. I have no objections (and frankly, nobody should even be asking me) about doing private builds.

**Comment by danmoseley** [comment]
I see. Bear in mind what we're contemplating would have to be on a public feed, since it will (obviously) need to be consumed by everyone contributing to this repo. It would have a distinct package name. Would that be acceptable?

**Comment by clairernovotny** [comment]
@danmoseley In order to justify the effort, I think it should just be a regular release to the main nuget.org feed so everyone can benefit, not just from this fix, but the reprodicuble build validation efforts.

**Comment by bradwilson** [comment]
>so everyone can benefit

Nobody uses this code but the .NET team. It's from `dotnet xunit` days, which was long ago deprecated, and the build of the console runner that consumes this code was only left in place because the .NET team asked me not to remove it.

**Comment by bradwilson** [comment]
>I am also seeing this intermittently on Mono runtime test lanes.

I realize I'm responding to an old comment here, but be aware that Mono has never been a supported platform for xUnit.net prior to v3.

**Comment by clairernovotny** [comment]
@bradwilson the other benefit I was referring to was to build with the latest toolchain & settings so that the libraries can pass the reproducible validation.

**Comment by bradwilson** [comment]
That benefit is orthogonal to whether it's a "private" build or a "sanctioned" one.

**Comment by clairernovotny** [comment]
Not entirely as that's the build "everyone else" will use too and once NuGet uses that in package scoring/health, it'll help.

**Comment by bradwilson** [comment]
If the .NET team decides this is an important part of their decision, then they will need to commit someone to being the point of contact for all v2 work. I will not accept a "one and done" official build and then ignoring the rest of the community. The .NET team does not get special treatment; if they want this to happen, then they need to commit to enabling it for everybody else, too (and we'll need to arrive on a process for those changes).

**Comment by kunalspathak** [comment]
Another one: https://dev.azure.com/dnceng/public/_build/results?buildId=1244680&view=ms.vss-test-web.build-test-results-tab&runId=36994542&resultId=109776&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab

**Comment by radical** [comment]
Another - https://dev.azure.com/dnceng/public/_build/results?buildId=1252182&view=logs&j=2296eecc-f5ab-523f-c15d-f45eef1bfdea&t=d328b091-66a6-5d96-5a21-3cf77f57f009

**Comment by ViktorHofer** [comment]
![image](https://user-images.githubusercontent.com/7412651/130212755-12ba306f-0fd2-4a48-9e68-d3f60837c2d6.png)

from https://runfo.azurewebsites.net/tracking/issue/111/. This is one of the top contributors to our test flakiness. Also ranked second in https://github.com/dotnet/runtime/issues/702.

@markwilkie @ChadNedzlek could we leverage test retries to avoid this issue? This is a long standing one and we haven't yet made any progress on it as the relevant code is owned by xunit and not maintained anymore.

This would have quite a big impact on our reliability numbers, presumably much more than enabling test retries for networking tests. What we would need to be cautious about is that we don't hide any other "Collection was modified" issues. That should be achievable by only enabling retries if the Exception comes form xunit:

```
Unhandled exception. System.InvalidOperationException: Collection was modified after the enumerator was instantiated.
   at System.Collections.Generic.Stack`1.Enumerator.MoveNext()
   at Xunit.Sdk.DisposalTracker.Dispose() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\DisposalTracker.cs:line 26
   at Xunit.Sdk.ExtensibilityPointFactory.Dispose() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\ExtensibilityPointFactory.cs:line 53
   at Xunit.Sdk.TestFramework.Dispose() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\TestFramework.cs:line 52
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__127_1(Object state)
```

**Comment by ChadNedzlek** [comment]
I believe this manifests as a test crash.  We don't retry those right now, because the retries are based on tests failing or not, and this crash causes there to be no test results, yes?

We could try adding a new facility for this, but the problem becomes there is no real way to report it, since there is no test result to report it to.  We could potentially just _ignore_ the problem completely though.

**Comment by ViktorHofer** [comment]
Still happening many times per week.

Console: https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-488292d291624bbbb3/Interop/1/console.41684910.log?sv=2019-07-07&se=2021-09-14T05%3A22%3A46Z&sr=c&sp=rl&sig=Jz0mMRM9DiHSkUyCLfo4GqOsduxouVWjBsaahHdAwcU%3D

Build: https://dev.azure.com/dnceng/public/_build/results?buildId=1316687&view=ms.vss-test-web.build-test-results-tab&runId=38731668&resultId=102472&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab

**Comment by maryamariyan** [comment]
@ericstj and I were looking at the linked test failures again and found an interesting pattern. All of the clr test failures shown here were using XUnitWrapper in the logs. 

e.g. seen in [here](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-58526-merge-b24a7a8d2ed8452fa8/PayloadGroup0/1/console.6b8ebd58.log?sv=2019-07-07&se=2021-09-21T22%3A42%3A49Z&sr=c&sp=rl&sig=S2Ujp2xxEZdV8Ttkil4MDK6VKICCq2b1z60EQb23LZo%3D):
```
Unhandled exception.   Discovering: readytorun.r2rdump.XUnitWrapper (method display = ClassAndMethod, method display options = None)
  Discovered:  readytorun.r2rdump.XUnitWrapper (found 1 test case)
  Starting:    readytorun.r2rdump.XUnitWrapper (parallel test collections = on, max threads = 4)
System.InvalidOperationException: Collection was modified after the enumerator was instantiated.
   at System.Collections.Generic.Stack`1.Enumerator.MoveNext()
   at Xunit.Sdk.DisposalTracker.Dispose() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\DisposalTracker.cs:line 26
   at Xunit.Sdk.ExtensibilityPointFactory.Dispose() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\ExtensibilityPointFactory.cs:line 53
   at Xunit.Sdk.TestFramework.Dispose() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\TestFramework.cs:line 52
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__127_1(Object state)
   at System.Threading.QueueUserWorkItemCallbackDefaultContext.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
   at System.Threading.Thread.StartCallback()
```

Are we doing anything unusual in that test infrastructure?

**Comment by ChadNedzlek** [comment]
I think this is just a bug in the XUnit execution engine, which I don't think is under active development anymore.  I don't think it's caused by anything in the actual testing.

**Comment by ericstj** [comment]
@maryamariyan is actually looking at [fixing that bug](https://github.com/xunit/xunit/pull/2375), and as part of that fix we'd like to describe why it's happening.  Xunit has a static API that will operate on this shared stack, so something is calling that static API from multiple threads.  We were going to trace a test exectuion to see why that happens from multiple threads.  To find a good test to trace we started looking at the existing repros and that's when we noticed this pattern.  It seemed to curious not to ignore.

**Comment by BruceForstall** [comment]
That's how the CoreCLR tests work: a set of "XUnitWrapper" assemblies are built with one class for each test with an xunit "Fact" for the test, which gets invoked by xunit, which then spawns an execution of the test. You can see the built wrapper source code in the artifacts, e.g., artifacts\tests\coreclr\windows.x64.Checked\TestWrappers\JIT.CodeGenBringUpTests\JIT.CodeGenBringUpTests.XUnitWrapper.cs

@trylek can probably explain the process more

**Comment by ericstj** [comment]
> built with one class for each test 
I see, so perhaps this pattern is appearing because we're stressing Xunit.  It's interesting that we pass many test assemblies to a single invocation.  That could be the reason this happens more often with these wrapper tests.

**Comment by BruceForstall** [comment]
Note that we pass `-parallel collections` to xunit. I think (?) that allows tests invoked by different wrapper assemblies to be run in parallel, but those in a single wrapper assembly must be serialized. I can't remember why we did that; possibly to avoid hitting this issue (as much).

**Comment by agocke** [comment]
I believe `-parallel collections` means to parallelize every XUnit collection, which is controlled by the XUnit `Collection` attribute that you can put on test classes. If a class isn't attributed, I think all classes are by-default different collections, so effectively every test class is executed in parallel.

**Comment by ericstj** [comment]
Interesting, that jives with "stressing" xunit.  Another characteristic we noticed was that the tests failing were all `PayloadGroup0`.  At first I thought this was just a reporting quirk, but now I noticed that it's actually the work item name: 
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/tests/Common/testgrouping.proj#L119-L137

**Comment by ViktorHofer** [comment]
> @ericstj and I were looking at the linked test failures again and found an interesting pattern. All of the clr test failures shown here were using XUnitWrapper in the logs.

Even though the failure occurs in runtime tests most of the time (which use the xunitwrapper infra) there are also hits in libraries tests.

**Comment by ericstj** [comment]
I'd be very interested in seeing where this hits in the libraries tests.  If you have any logs please share them.  I couldn't find any in the linked builds.

After looking at the call-stacks here and the tests they repro in, it's all happening due to us running multiple test assemblies at once.  The `TestFramework` object has a 1:1 relationship with test assembly and when that object is disposed it calls the static ExtensibilityFactory.Dispose which modifies a static DisposalTracker field without a lock.  These runtime tests use a *lot* of assemblies with very few tests, so they have a high likelihood of concurrent execution.  @bradwilson mentioned that .NETCore xunit runner isn't designed to handle running multiple assemblies at once and we might be the only folks doing it due to our special runner.  It might be a good idea to avoid multiple assemblies long term, as it puts us off the happy path and might cause other issues.  I can see us dealing with that by either merging the assemblies (they are all generated after all) or making one call per assembly.

We'll still pursue the xunit fix, but I wanted to point out we are doing something rather special here that seems to be causing/exacerbating this problem.

**Comment by agocke** [comment]
>  @bradwilson mentioned that .NETCore xunit runner isn't designed to handle running multiple assemblies at once

I don't understand this. https://xunit.net/docs/running-tests-in-parallel seems to document the option as the default for the XUnit 1 and 2 console runners.

**Comment by ericstj** [comment]
Test cases work great in parallel, but multiple assemblies in parallel is not supported.  Apparently it's only working for us because we have a fork of the runner.

**Comment by agocke** [comment]
Well, `assemblies` is a documented option to the runner, it's just not the default. Maybe it's just less robust?

**Comment by ViktorHofer** [comment]
> Apparently it's only working for us because we have a fork of the runner.

Our fork is exactly the same as the upstream xunit.console runner (unsure if it still exists upstream) except for the following changes which we documented here: https://github.com/dotnet/arcade/blob/main/src/Microsoft.DotNet.XUnitConsoleRunner/README.md.

**Comment by ericstj** [comment]
My understanding is that even the upstream runner didn't support multiple test assemblies very well and was a motivating factor for a change in direction in v3.  V3's console runner only supports .NETFramework.  I believe the runner was in this state because it was originally designed for .NETFramework which provided AppDomain isolation and test-specific dependency control that could allow for isolation of runner:test and test:test dependencies.  We didn't have parity for that in .NETCore.  Now we have some support with ALC functionality but I suspect it's not equivalent and I believe the decision was already made to change direction.

> Maybe it's just less robust?

Yes, though the 2.0 runner supports that commandline option, we might be the only ones using it, and likely are the only ones stressing it to the extent we do in dotnet/runtime.  So if we could refactor to single-assembly per invocation we'd be more likely to be on the happy path and avoid these sort of bugs.

**Comment by bradwilson** [comment]
The issue (in v2) is this:

We support running multiple assemblies in parallel, but the only runners that officially support this are the .NET Framework runners (our console and MSBuild runners). There is app domain isolation available there that helps prevent multiple test assemblies from stepping on each other, though there are edge cases inside .NET Framework itself that can still confound (like the fact that "current directory" is a process-wide setting, so calling `Directory.SetCurrentDirectory` affects all app domains in the process. But by and large, app domain isolation does all the isolation heavy lifting).

Most teams--I want to say "everybody _except_ the .NET team" but I don't have the data to back that up--who run .NET Core unit tests use the third party `dotnet test` functionality (or something that consumes that same plugin, like Visual Studio and VS Code). Whether those things support running multiple assemblies in parallel is up to them, and how they deal with isolation is up to them (the only reasonable option they'd have is process isolation, I assume). We don't provide that functionality.

What the .NET team appears to be doing is using a forked version of a deprecated runner that was a .NET Core version of our console runner (originally designed for the now removed `dotnet xunit` functionality), which is attempting to run multiple assemblies in the same address space. .NET Core does not have app domain isolation (or any other isolation, for that matter) that can help prevent the kinds of collisions that you've hit. This is why the bug your team is hitting is only being hit by you: because you're the only people trying to write your own runner in this way (or if anybody else is doing so, they've fixed or ignored the problem in their one way without raising it with xUnit.net).

As for v3, the entire design of test projects has changed from "compiles to library" to "compiles to application". The issues with isolation and dependency resolution in .NET Core were a big part of the forcing function. In v3, our supported console runner will allow you to run .NET Core tests in addition to .NET Framework tests (and even parallelize .NET Framework and .NET Core tests together, as well as having significantly better support for varying versions of .NET Framework or .NET Core, allow running 32- and 64-bit tests in parallel, etc.).

**Comment by ViktorHofer** [comment]
According to runfo this isn't happening anymore in main or release/6.0: https://runfo.azurewebsites.net/tracking/issue/111/

As the fix was merged into release/6.0 but not release/6.0-rc2 there are still a few hits for the RC2 branch but those will just naturally go away when RC2 doesn't produce any builds anymore.

Closing.
